File: hosts/ios/App/Source/Gateway/GatewayCore.swift
import Foundation

/// One settled primitive call. `payload` is JSON-serializable (a dictionary,
/// an array, a String, a number, or NSNull for void/null results).
typealias GatewayDone = (Result<Any, GatewayError>) -> Void

/// Handler signature: runs OFF the runtime thread; must call `done` exactly
/// once (from any thread — the core routes the settle back onto the runtime
/// queue, ARCHITECTURE.md §6 thread rules).
typealias GatewayHandler = (GatewayCall, @escaping GatewayDone) -> Void

/// Arguments of one primitive call, decoded from the bridge's args JSON.
/// Bytes ride base64 in fields ending "B64" (frozen bridge convention).
struct GatewayCall {
    let callId: Int
    let args: [String: Any]

    func string(_ key: String) -> String? { args[key] as? String }

    func bool(_ key: String, default def: Bool = false) -> Bool {
        args[key] as? Bool ?? def
    }

    func dict(_ key: String) -> [String: Any] { args[key] as? [String: Any] ?? [:] }

    /// Decodes a base64 byte field ("<key>B64"); nil when absent/malformed.
    func b64(_ key: String) -> Data? {
        guard let text = string(key) else { return nil }
        return Data(base64Encoded: text)
    }
}

/// Contract §3 GatewayError — one structured rejection per failed call.
/// Codes are the closed set of primitives.md (unknown codes are fatal).
struct GatewayError: Error {
    let code: String
    let primitive: String
    let message: String
}

/// The caller's manifest (data-protocols.md §2), read from the staged bundle
/// root. Fail-loud: a missing or malformed manifest aborts the session.
struct GatewayManifest {
    static let caller = "dsh.spike.scenario"
    let id: String
    let required: [String]

    init(bundleRoot: URL) throws {
        let url = bundleRoot.appendingPathComponent("manifest.json")
        guard let data = FileManager.default.contents(atPath: url.path),
              let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        else {
            throw GatewayError(code: "io", primitive: "manifest",
                message: "bundle_root/manifest.json missing or unparseable")
        }
        guard obj["schemaVersion"] as? Int == 1, let id = obj["id"] as? String,
              let caps = obj["capabilities"] as? [String: Any],
              let required = caps["required"] as? [String]
        else {
            throw GatewayError(code: "io", primitive: "manifest",
                message: "manifest.json violates schemaVersion 1 shape")
        }
        self.id = id
        self.required = required
        guard id == Self.caller else {
            throw GatewayError(code: "io", primitive: "manifest",
                message: "unexpected caller identity \(id)")
        }
    }

    /// Permission-flag grammar: `<name>` or `<name>@<major>` — the caller
    /// holds the flag when any required entry names the primitive exactly or
    /// with a major suffix.
    func grants(primitive: String) -> Bool {
        required.contains { $0 == primitive || $0.hasPrefix(primitive + "@") }
    }
}

/// The capability gateway core: dispatch table name→handler, permission
/// enforcement against the manifest, and the mandatory structured audit of
/// primitives.md §6 — one record per call, never payload contents. Audit
/// lines use their own stdout prefix ("dsh.gateway.audit: ") so the
/// canonical "dsh.spike.log: " E2E stream stays one-to-one.
final class GatewayCore {
    /// The nine frozen primitives (contract/primitives.md §2).
    static let primitives = [
        "fsRead", "fsWrite", "fsScope", "httpFetch", "notify",
        "presentApproval", "presentPicker", "keychainGet", "keychainSet",
    ]
    static let auditPrefix = "dsh.gateway.audit: "

    let manifest: GatewayManifest
    private let workQueue = DispatchQueue(label: "org.dsh.gateway.work")
    private var handlers: [String: GatewayHandler] = [:]
    private let auditLock = NSLock()

    /// Wired by the session; each hops onto the runtime thread (the only
    /// thread allowed to touch the C runtime).
    var settle: ((Int, Bool, String) -> Void)?
    var emit: ((String) -> Void)?

    init(bundleRoot: URL) throws {
        manifest = try GatewayManifest(bundleRoot: bundleRoot)
    }

    func register(name: String, _ handler: @escaping GatewayHandler) {
        handlers[name] = handler
    }

    /// Entry point of the frozen bridge's on_call — invoked ON THE RUNTIME
    /// THREAD. Unknown or ungranted primitives settle denied with a "denied"
    /// audit verdict; granted calls run their handler off-thread.
    func dispatch(callId: Int, name: String, argsJSON: String) {
        if name == "httpFetch.abort" { return dispatchAbort(callId, argsJSON) }
        let base = name.split(separator: ".").first.map(String.init) ?? name
        guard let handler = handlers[name], manifest.grants(primitive: base) else {
            audit(primitive: name, verdict: "denied", outcome: "denied")
            let message = "primitive not granted to \(manifest.id)"
            let error = GatewayError(code: "denied", primitive: name, message: message)
            settle?(callId, false, Self.errorJSON(error))
            return
        }
        guard let data = argsJSON.data(using: .utf8),
              let args = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        else {
            audit(primitive: name, verdict: "granted", outcome: "invalid")
            let error = GatewayError(
                code: "invalid", primitive: name, message: "args JSON unparseable")
            settle?(callId, false, Self.errorJSON(error))
            return
        }
        let call = GatewayCall(callId: callId, args: args)
        dispatchGranted(callId: callId, name: name, handler: handler, args: call)
    }

    /// Runs the handler on the work queue; the completion re-enters here for
    /// audit + the settle hop back onto the runtime queue.
    private func dispatchGranted(
        callId: Int, name: String, handler: @escaping GatewayHandler,
        args: GatewayCall
    ) {
        let done: GatewayDone = { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let payload):
                self.audit(primitive: name, verdict: "granted", outcome: "ok")
                self.settle?(callId, true, Self.encode(payload))
            case .failure(let error):
                self.audit(primitive: name, verdict: "granted", outcome: error.code)
                self.settle?(callId, false, Self.errorJSON(error))
            }
        }
        workQueue.async { handler(args, done) }
    }

    /// Control-plane call of the frozen bridge: JS global __dshGatewayAbort
    /// arrives as on_call with name "httpFetch.abort", args {"callId":N}.
    /// The handler parses the target (number or "body:<callId>" string).
    /// Fire-and-forget: the C side assigns abort on_calls no pending call
    /// id, so — unlike primitives — they are audited but never settled.
    private func dispatchAbort(_ callId: Int, _ argsJSON: String) {
        let args = (try? JSONSerialization.jsonObject(with: Data(argsJSON.utf8)))
            as? [String: Any]
        handlers["httpFetch.abort"]?(GatewayCall(callId: -1, args: args ?? [:]), { _ in })
        audit(primitive: "httpFetch.abort", verdict: "granted", outcome: "ok")
    }

    // ---- audit (mandatory, host-fixed, never payload contents) -------------

    func audit(primitive: String, verdict: String, outcome: String) {
        let record: [String: Any] = [
            "ts": Self.isoNow(),
            "primitive": primitive,
            "caller": manifest.id,
            "verdict": verdict,
            "outcome": outcome,
        ]
        guard let line = Self.jsonLine(record) else { return }
        auditLock.lock()
        print(Self.auditPrefix + line)
        fflush(stdout)
        NSLog("%@", Self.auditPrefix + line)
        auditLock.unlock()
    }

    // ---- shared JSON helpers ------------------------------------------------

    static func errorJSON(_ error: GatewayError) -> String {
        let obj: [String: Any] = [
            "code": error.code, "primitive": error.primitive, "message": error.message,
        ]
        return jsonLine(obj) ?? "{}"
    }

    static func encode(_ payload: Any) -> String { jsonLine(payload) ?? "null" }

    static func jsonLine(_ obj: Any) -> String? {
        guard JSONSerialization.isValidJSONObject(obj) || obj is NSNull else { return nil }
        if obj is NSNull { return "null" }
        guard let data = try? JSONSerialization.data(withJSONObject: obj) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func isoNow(_ date: Date = Date()) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.string(from: date)
    }

    /// stdout automation marker (NOT the canonical stream) for the E2E
    /// driver: "spike: ui-wait <name>" before an automatable surface,
    /// "spike: ui-done <name>" once it resolves.
    static func uiMarker(_ name: String, _ phase: String) {
        print("spike: ui-\(phase) \(name)")
        fflush(stdout)
    }
}

File: hosts/android/app/src/main/java/com/dshmobile/spike/GatewayCore.kt
package com.dshmobile.spike

import android.util.Log
import org.json.JSONObject

/**
 * The capability gateway core — Kotlin sibling of hosts/ios Gateway/-
 * GatewayCore.swift. Dispatch table name→handler, permission enforcement
 * against the caller's bundle manifest (data-protocols.md §2), and the
 * mandatory structured audit of contract/primitives.md §6 — one record per
 * call, never payload contents. Audit rides its own logcat tag
 * ("dsh.spike.audit") with the flat "dsh.gateway.audit: " prefix so the
 * canonical "dsh.spike.log: " E2E stream stays one-to-one. Handlers run on
 * the dispatching (runtime) thread and call done OFF it — every settle hop
 * back via SpikeRuntime.post (ARCHITECTURE.md §6 thread rules).
 */
class GatewayCore private constructor(val manifest: GatewayManifest) {

    companion object {
        /** The caller identity of the spike bundle (manifest.json id). */
        const val CALLER = "dsh.spike.scenario"

        /** The nine frozen primitives (contract/primitives.md §2). */
        val PRIMITIVES = listOf(
            "fsRead", "fsWrite", "fsScope", "httpFetch", "notify",
            "presentApproval", "presentPicker", "keychainGet", "keychainSet",
        )
        const val AUDIT_PREFIX = "dsh.gateway.audit: "
        private const val AUDIT_TAG = "dsh.spike.audit"
        private const val UI_TAG = "dsh.spike.ui"

        /** Fails loud (rules.md rule 5): bundle_root/manifest.json must exist. */
        fun create(bundleRoot: java.io.File): GatewayCore {
            val file = java.io.File(bundleRoot, "manifest.json")
            val obj = JSONObject(file.readText())
            val id = obj.getString("id")
            if (id != CALLER) {
                error("unexpected caller identity $id")
            }
            val required = obj.getJSONObject("capabilities").getJSONArray("required")
                .let { arr -> List(arr.length()) { arr.getString(it) } }
            return GatewayCore(GatewayManifest(id, required))
        }

        /** Canonical JSON line for a GatewayError-shaped rejection. */
        fun errorJSON(code: String, primitive: String, message: String): String =
            JSONObject().put("code", code).put("primitive", primitive)
                .put("message", message).toString()

        /** E2E driver marker (NOT the canonical stream): ui-wait / ui-done. */
        fun uiMarker(name: String, phase: String) {
            Log.i(UI_TAG, "ui-$phase $name")
        }
    }

    class GatewayError(
        val code: String,
        val primitive: String,
        message: String,
    ) : Exception(message)

    /** The caller identity + its required capability strings. */
    class GatewayManifest(val id: String, val required: List<String>) {
        /** `<name>` or `<name>@<major>` grammar (contract §6). */
        fun grants(primitive: String): Boolean =
            required.any { it == primitive || it.startsWith("$primitive@") }
    }

    /** One settled primitive call; args is the bridge's parsed JSON object. */
    class GatewayCall(val callId: Int, val args: JSONObject) {
        fun string(key: String): String? =
            if (args.has(key) && !args.isNull(key)) args.getString(key) else null

        fun optLong(key: String): Long? =
            if (args.has(key) && !args.isNull(key)) args.getLong(key) else null
    }

    fun interface Done {
        /** Payload = JSON value (JSONObject/Boolean/String/number/null). */
        fun settle(result: Any?, error: GatewayError?)
    }

    private val handlers = HashMap<String, (GatewayCall, Done) -> Unit>()

    /** Wired by the session; hops onto the runtime thread (JNI). */
    var settleFn: ((callId: Int, ok: Boolean, json: String) -> Unit)? = null

    fun register(name: String, handler: (GatewayCall, Done) -> Unit) {
        handlers[name] = handler
    }

    /** Entry point of the frozen bridge's on_call — RUNTIME THREAD. Unknown
     * or ungranted primitives settle denied with a "denied" audit verdict;
     * granted calls run their handler (which settles off-thread). */
    fun dispatch(callId: Int, name: String, argsJSON: String) {
        if (name == "httpFetch.abort") return dispatchAbort(name, argsJSON)
        val base = name.substringBefore('.')
        val handler = handlers[name]
        if (handler == null || !manifest.grants(base)) {
            audit(name, "denied", "denied")
            settleFn?.invoke(
                callId, false,
                errorJSON("denied", name, "primitive not granted to ${manifest.id}"),
            )
            return
        }
        val args = try {
            JSONObject(if (argsJSON.isBlank()) "{}" else argsJSON)
        } catch (e: Exception) {
            audit(name, "granted", "invalid")
            settleFn?.invoke(callId, false, errorJSON("invalid", name, "args JSON unparseable"))
            return
        }
        val done = Done { result, error ->
            if (error != null) {
                audit(name, "granted", error.code)
                settleFn?.invoke(callId, false, errorJSON(error.code, name, error.message ?: ""))
            } else {
                audit(name, "granted", "ok")
                val json = when (result) {
                    null -> "null"
                    is JSONObject -> result.toString()
                    is String -> JSONObject.quote(result)
                    is Boolean, is Int, is Long -> result.toString()
                    else -> error("unsupported settle payload ${result::class.java}")
                }
                settleFn?.invoke(callId, true, json)
            }
        }
        try {
            handler(GatewayCall(callId, args), done)
        } catch (e: Exception) {
            done.settle(
                null,
                GatewayError("io", name, "${e::class.java.simpleName}: ${e.message ?: ""}"),
            )
        }
    }

    /** Control-plane abort (frozen bridge): audited, never settled. */
    private fun dispatchAbort(name: String, argsJSON: String) {
        val call = GatewayCall(-1, try {
            JSONObject(if (argsJSON.isBlank()) "{}" else argsJSON)
        } catch (_: Exception) {
            JSONObject()
        })
        handlers[name]?.invoke(call) { _, _ -> }
        audit(name, "granted", "ok")
    }

    /** Mandatory structured audit (contract §6): never payload contents. */
    private fun audit(primitive: String, verdict: String, outcome: String) {
        val record = JSONObject()
            .put("ts", java.time.Instant.now().toString())
            .put("primitive", primitive)
            .put("caller", manifest.id)
            .put("verdict", verdict)
            .put("outcome", outcome)
        Log.i(AUDIT_TAG, AUDIT_PREFIX + record)
    }
}
